\documentclass[webpdf,contemporary,large,namedate]{oup-authoring-template}%

%\PassOptionsToPackage{hyphens}{url}
%\PassOptionsToPackage{colorlinks,linkcolor=blue,urlcolor=blue,citecolor=blue,anchorcolor=blue}{hyperref}

\DeclareMathOperator*{\argmax}{argmax}

\usepackage{algorithmicx}
\usepackage{lmodern}
\usepackage{setspace}
\renewcommand{\ttdefault}{cmtt}

\begin{document}
\journaltitle{TBD}
\DOI{TBD}
\copyrightyear{2025}
\pubyear{2025}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Preprint}
\firstpage{1}

\title[Improving spliced alignment]{Modeling splice sites with deep learning Improves spliced alignment}
\author[1]{Siying Yang}
\author[1,2]{Neng Huang}
\author[1,2,3,$\ast$]{Heng Li\ORCID{0000-0003-4874-2874}}
\address[1]{Department of Biomedical Informatics, Harvard Medical School, 10 Shattuck St, Boston, MA 02215, USA}
\address[2]{Department of Data Science, Dana-Farber Cancer Institute, 450 Brookline Ave, Boston, MA 02215, USA}
\address[3]{Broad Insitute of MIT and Harvard, 415 Main St, Cambridge, MA 02142, USA}
\corresp[$\ast$]{Corresponding author. \href{mailto:hli@ds.dfci.harvard.edu}{hli@ds.dfci.harvard.edu}}

%\received{Date}{0}{Year}
%\revised{Date}{0}{Year}
%\accepted{Date}{0}{Year}

\abstract{
\sffamily\footnotesize
\textbf{Motivation:}
Spliced alignment refers to the alignment of messenger RNA (mRNA) or protein sequences to eukaryotic genomes.
It plays a critical role in gene annotation and the study of gene functions.
Accurate spliced alignment demands sophisticated modeling of splice sites,
but current aligners use simple models, which may affect their accuracy given dissimilar sequences.
\vspace{0.5em}\\
\textbf{Results:}
We implemented minisplice that learns splice signals with one-dimensional convolutional neural network (1D-CNN)
and estimates the empirical splicing probability for every {\tt GT} and {\tt AG} sites in the genome.
We modified minimap2 and miniprot to take advantage of pre-computed splicing probability during alignment.
We trained a model for vetebrates and insect.
Applied to human long-read RNA-seq data and cross-species protein datasets,
the model greatly improves the junction accuracy especially for noisy long reads
or proteins of distant homology.
\vspace{0.5em}\\
\textbf{Availability and implementation:}
\url{https://github.com/lh3/minisplice}
}

\maketitle

\section{Introduction}

RNA splicing is the process of removing introns from precursor mRNAs (pre-mRNAs).
It is widespread in eukaryotic genomes~\citep{Keren:2010aa}.
%The 5'-end of an intron is called a \emph{donor} site
%and the 3'-end called an \emph{acceptor} site.
In human, for example, each protein-coding gene contains 9.4 introns on average;
$>$98\% of introns start with {\tt GT} on the genome (or more precisely {\tt GU} on the pre-mRNA)
and $>$99\% end with {\tt AG}~\citep{Sibley:2016vh}.
On the other hand, there are hundreds of millions of di-nucleotide {\tt GT} or {\tt AG}
in the human genome.
Only $\sim$0.1\% of them are real splice sites.
Identifing real splice sites, which is required for gene annotaiton,
is challenging due to the low signal-to-noise ratio.

To annotate splice sites and genes in a new genome,
we can perform RNA sequencing (RNA-seq) and align mRNA sequences to the target genome.
This approach does not work well for genes lowly expressed in tissues being sequenced.
A complement strategy is to align mRNA or protein sequences from other species to the target genome.
Spliced alignment through introns is essential in both cases.

It is important to look for splice signals during spliced alignment
as the residue alignment around a splice site can be ambiguous.
For example, the three alignments in Fig.~\ref{fig:1} are equally good if we ignore splice signals.
However, as the putative intron in alignment (1) does not match the {\tt GT..AG}
signal, it is unlikely to be real.
While both (2) and (3) match the signal,
alignment (3) is more probable because it fits the splice consensus {\tt GTR..YAG} better~\citep{Irimia:2008aa,Iwata:2011aa},
where ``{\tt R}'' stands for an {\tt A} or a {\tt G} base and ``{\tt Y}'' for {\tt C} or {\tt T}.
In this toy example, the query sequence matches the reference perfectly in all three cases.
On real data, spliced aligners may introduce extra mismatches and gaps to reach splice sites.
The splice model has a major influence on the final alignment especially for diverged seuqences
when aligners need to choose between multiple similarly scored alignments around splice junctions.

\begin{figure}[b]
\centering
\includegraphics[width=.35\textwidth]{fig1}
\caption{Ambiguity in spliced alignment.
The same genome-mRNA sequence pair can be aligned differently without mismatches or gaps.}\label{fig:1}
\end{figure}

Position weight matrix (PWM) is a classical method for modeling splice signals~\citep{Staden:1984aa}.
It however does not perform well because it cannot capture dependencies between positions~\citep{Burge:1997uu}
or model regulatory motifs that do not have fixed positions.
Many models have been developed to overcome the limitation of PWM~\citep{Capitanchik:2025aa}.
In recent years, deep learning is gaining attraction
and has been shown to outperform traditional methods~\citep{Zhang:2018aa,DBLP:journals/access/DuYDZZL18,Albaradei:2020aa}.
Early deep learning models are small with only a few 1D-CNN layers.
Later models are larger, composed of residual blocks~\citep{Jaganathan:2019aa,Zeng:2022aa,Xu:2024aa,Chao:2024aa} or transformer blocks~\citep{You:2024aa,Chen:2024aa}.
It is also possible to fine tune genomic large-language models for splice site prediction~\citep{Nguyen:2023aa,Dalla-Torre:2025aa,Brixi2025.02.18.638918}.
Developed for general purposes, large-language models are computationally demanding and may be overkilling if we just use them to predict splice sites.

While qualified spliced aligners all look for the {\tt GT..AG} splice signal,
they model additional flanking sequences differently.
Intra-species mRNA-to-genome aligners such as BLAT~\citep{Kent:2002jk}, GMAP~\citep{Wu:2005vn} and Splign~\citep{Kapustin:2008tq} often do not model extra sequences beyond {\tt GT..AG}
because alignment itself provides strong evidence and ambiguity shown in Fig~\ref{fig:1} is rare.
Minimap2~\citep{Li:2018ab} prefers the {\tt GTR..YAG} consensus~\citep{Irimia:2008aa}.
This helps to improve the alignment of noisy long RNA-seq reads.
Protein-to-genome aligners tend to employ better models due to more ambiguous alignment given distant homologs.
Miniprot~\citep{Li:2023ab} considers rarer {\tt GC..AG} and {\tt AT..AC} splice sites and optionally prioritizes on the {\tt G|GTR..YNYAG} consensus
common in vertebrate and insect~\citep{Iwata:2011aa}, where ``{\tt |}'' indicates splice boundaries.
Exonerate~\citep{Slater:2005aa}, Spaln~\citep{Gotoh:2008aa,Iwata:2012aa,Gotoh:2024aa} and the original GeneWise~\citep{Birney:2004uy} use PWM.
GeneSeqer~\citep{Usuka:2000vi} and GenomeThreader~\citep{DBLP:journals/infsof/GremmeBSK05} apply more advanced models~\citep{Brendel:1998aa,Brendel:2004aa}.
Deep learning models have been applied to refining splice sites as a post-processing step~\citep{Chao:2024aa,Xia:2023aa}
but have not been integrated into spliced aligners.

In this article, we introduce minisplice, a command-line tool implemented in C,
that learns splice signals and scores candidate splice sites with a small 1D-CNN model with 35 thousand parameters.
We have modified minimap2 and miniprot to take the splice scores as input for improved spliced alignment.
Importantly, we aim to develop a simple model that is more capable than PWM and is still easy to deploy;
we do not intend to compete with the best splice models which are orders of magnitude larger.

\section{Methods}

\begin{figure}[b]
\includegraphics[width=.48\textwidth]{fig2}
\caption{Method overview. {\bf (a)} Overall workflow.
Training and prediction are done by minisplice.
Alignment is done by minimap2 or miniprot.
{\bf (b)} Default model architecture and parameterization.
Shaded boxes contain free parameters.}\label{fig:wf}
\end{figure}

Our overall workflow consists of three steps: training, prediction and alignment (Fig.~\ref{fig:wf}a).
First, we train a deep learning model and transform scores outputted by the model
to empirical probabilities using known gene annotation.
Second, given a target genome mRNA or protein sequences are aligned to,
we predict the empirical probability of splice sites at each {\tt GT} or {\tt AG} in the genome
and output the logarithm-scaled splice scores to a file.
Third, when aligning mRNA sequences with minimap2~\citep{Li:2018ab} or aligning protein sequences with miniprot~\citep{Li:2023ab},
we feed the precomputed splice scores to the aligners which use the scores during dynamic-programming-based residue alignment.
This procedure will improve the accuracy around splice sites.
As we will show later, we can merge the training data from several vertebrates and insect
to obtain a model working well in large phylogenetic clades.
We do not need to apply training often.

\subsection{Generating training data}

Consistent with the literature, we call the 5'-end of an intron as the \emph{donor} site
and 3'-end as the \emph{acceptor} site.
Minisplice takes the genome sequence in the FASTA format
and gene annotation in the 12-column BED format (BED12) as input.
It provides a script to convert annotation in GTF or GFF3 format to BED12.
To generate training data, minisplice inspects each annotated donor site and labels it as a positive site if its sequence is {\tt GT};
similarly, minisplice labels a positive acceptor if it is an annotated {\tt AG}.
We ignore donor sites without {\tt GT} or acceptor sites without {\tt AG}
because other types of splice sites are rare~\citep{Sibley:2016vh}.
The alignment step still considers rarer {\tt GC..AG} or {\tt AT..AC} splicing.

Due to potentially incomplete gene annotation in non-model organisms,
a {\tt GT} dinucleotide could be a real unannotated donor site.
To alleviate this issue, we label an unannotated {\tt GT} as a negative donor site
only if it comes from the opposite strand of an annotated gene~\citep{Chao:2024aa}.
In the rare case when two genes on opposite strands overlap with each other,
we ignore the overlapping region.
To balance positive and negative sites,
we downsample negative sites such that the positive-to-negative ratio is 1:3.
For each positive or negative donor site,
we extract 100bp immediately before and after {\tt GT}.
The total length of sequences used for training is thus 202bp.
Negative acceptor sites are generated similarly.

\subsection{Model architecture}

Minisplice uses a model with two 1D-CNN layers (Fig.~\ref{fig:wf}b).
The architecture is common among small models for splice site prediction~\citep{Zabardast:2023aa}.
The default model uses 16 features at both CNN layers and has 34,978 free parameters in total (sum of numbers in shaded boxes).
We briefly evaluated alternative models, for example with different kernel sizes, more CNN or dense layers, more dropout layers or more parameters.
We observed comparable validation cost with similar number of parameters.
We chose a relatively small model in the end because it is more efficient to apply.

\subsection{Training and testing}

We use 80\% of genes on the odd chromosomes or contigs for training and reserve the rest 20\% for validation.
Recall that we intend to predict splice sites across the whole genome
but when generating training data, we downsample negative {\tt GT}/{\tt AG} to a small fraction.
To test the model accuracy in a setting closer to the prediction task,
we apply the trained model to every {\tt GT}/{\tt AG} on the even chromosomes or contigs
and compare the prediction to the known gene annotation.
In comparision to training data, testing data is more correlated with the quality of known gene annotation:
missing junctions in the annotation would appear to be false positives (FPs),
while falsely annotated junctions would look false negatives (FNs).
Testing accuracy is a potential indicator of gene annotation quality.

\subsection{Transforming raw model scores to proabilities}

With the `softmax' operator at the end,
the model scores each candidate splice site with a number between 0 and 1, the higher the better.
This score is not probability in particular when the property of the training data
is distinct from our intended application.
We need to transform this score to probability to work with the probability-based scoring system of miniprot.

We evenly divide raw model scores into $b$ bins (50 by default).
Raw score $t\in[0,1)$ would be assigned to bin $i=\lfloor tb\rfloor$.
Let $P_i$ be the number of annotated splice sites scored to bin $i$, $i=0,1,\ldots,b-1$,
and $N_i$ be the number of unannotated {\tt GT}/{\tt AG} sites scored to bin $i$.
$P_i/(P_i+N_i)$ is the empirical probability of a candidate site in bin $i$ being real.
Let $P=\sum_i P_i$ and $N=\sum_i N_i$.
Given raw score $t$, the transformed score is
\begin{equation}\label{eq:s}
s(t)\triangleq 2\log_2\left(\frac{P_{\lfloor tb\rfloor}}{P_{\lfloor tb\rfloor}+N_{\lfloor tb\rfloor}}\cdot\frac{P+N}{P}\right)
\end{equation}
It computes the log odds of the probability estimiated with the deep learning model
versus with the null model that assumes every {\tt GT}/{\tt AG} having equal probability of being real.
The $2\log_2$ scaling comes from BLOSUM scoring matrices~\citep{Henikoff:1992tk} which miniprot uses.

\subsection{Aligner integration}

Minimap2~\citep{Li:2018ab} uses the following equation for spliced alignment:
\begin{equation}\label{eq:splice}
\left\{\begin{array}{l}
H_{ij} = \max\{H_{i-1,j-1}+s(i,j),E_{ij},F_{ij},\tilde{E}_{ij}-a(i)\}\\
E_{i+1,j}= \max\{H_{ij}-q,E_{ij}\}-e\\
F_{i,j+1}= \max\{H_{ij}-q,F_{ij}\}-e\\
\tilde{E}_{i+1,j}= \max\{H_{ij}-d(i)-\tilde{q},\tilde{E}_{ij}\}\\
\end{array}\right.
\end{equation}
where $q$ is the gap open penalty, $e$ the gap extension penalty
and $s(i,j)$ gives the substitution score between the $i$-th position
on the reference and the $j$-th position on the query sequence.
$d(i)$ and $a(i)$ are the donor and acceptor scores, respectively, calculated with Eq.~(\ref{eq:s}).
Miniprot~\citep{Li:2023ab} uses a more complex equation which has the same donor and acceptor score functions.

\subsection{Implementation}

Minisplice is implemented in the C programming language with the only dependency
being zlib for reading compressed files.
It uses a deep-learning library we developed earlier for identifying human centromeric repeats~\citep{Li:2019aa}.
Minisplice outputs splice scores in a TAB-delimited format like:
\begin{verbatim}
chr2   4184146   +   A   9
chr2   4184167   +   A   -5
chr2   4184191   -   D   5
chr2   4184199   +   A   -5
chr2   4184213   +   D   3
chr2   4184219   -   A   1
\end{verbatim}
where the second column corresponds to the offset of the splice boundary
and the last column gives the splice score.
We modified minimap2 and miniprot to optionally take such a file as input
and use the splice scores during residue alignment.
Notably, minimap2 and miniprot do not directly depend on minisplice.
Users can provide splice scores estimated by other means in principle.

%The purple line in Fig.~\ref{fig:3} shows $s(t)$ estimated from data;
%the green line corresponds to $2\log_2\left[t\cdot(P+N)/P\right]$
%which is the score assuming raw score $t$ is a proper probability.
%
%\begin{figure}[tb]
%\centering
%\includegraphics[width=.4\textwidth]{fig3}
%\caption{Transformed splice score as a function of raw model score.
%0.0014 is the fraction of annotated splice sites out of all {\tt GT}/{\tt AG}
%genome-wide.}\label{fig:3}
%\end{figure}

\section{Results}

We evaluated the accuracy of trained models using Receiver Operating Characteristic (ROC) curves
where we computed the true positive rate and false positive rate at different thresholds on raw model scores.
To find a small model that is generalized to multiple species and is fast to deploy,
we experiemented models under a variety of settings.
Our final model is trained on six insect from five orders and seven vertebrates.
We did not train a plant model because we are less familiar with the plant phylogeny.

\begin{table}[!tb]
\caption{Datasets\label{tab:data}}
\begin{tabular*}{\columnwidth}{@{\extracolsep\fill}lll@{\extracolsep\fill}}
\toprule
Label & Species & Accession \\
\midrule
human\dag      & \emph{Homo sapiens}               & GCA\_000001405.29 \\
mouse\dag*     & \emph{Mus musculus}               & GCA\_000001635.9 \\
chicken\dag*   & \emph{Gallus gallus}              & GCA\_016699485.1 \\
zebrafish\dag* & \emph{Danio rerio}                & GCA\_000002035.4 \\
fruitfly\dag*  & \emph{Drosophila melanogaster}    & GCA\_000001215.4 \\
mosquito\dag   & \emph{Anopheles gambiae}          & GCA\_943734735.2 \\
mCanLup*       & \emph{Canis lupus baileyi}        & GCF\_048164855.1 \\
mLagAlb*       & \emph{Lagenorhynchus albirostris} & GCF\_949774975.1 \\
bAccGen*       & \emph{Astur gentilis}             & GCF\_929443795.1 \\
bAnaAcu        & \emph{Anas acuta}                 & GCF\_963932015.1 \\
aDenEbr        & \emph{Dendropsophus ebraccatus}   & GCF\_027789765.1 \\
fCarCar*       & \emph{Carassius carassius}        & GCF\_963082965.1 \\
fPunPun        & \emph{Pungitius pungitius}        & GCF\_949316345.1 \\
sMobHyp        & \emph{Mobula hypostoma}           & GCF\_963921235.1 \\
icTenMoli*     & \emph{Tenebrio molitor}           & GCF\_963966145.1 \\
idCalVici*     & \emph{Calliphora vicina}          & GCF\_958450345.1 \\
idStoCalc      & \emph{Stomoxys calcitrans}        & GCF\_963082655.1 \\
ihPlaCitr*     & \emph{Planococcus citri}          & GCF\_950023065.1 \\
ilCydFagi*     & \emph{Cydia fagiglandana}         & GCF\_963556715.1 \\
ilOstNubi      & \emph{Ostrinia nubilalis}         & GCF\_963855985.1 \\
iyBomTerr*     & \emph{Bombus terrestris}          & GCF\_910591885.1 \\
iyVesCrab      & \emph{Vespa crabro}               & GCF\_910589235.1 \\
\botrule
\end{tabular*}
\begin{tablenotes}\setlength\itemsep{0.0em}
Ensembl or Gencode annotations were used for model organisms (marked by ``\dag'');
RefSeq annotations for non-model organisms whose labels
follow the naming standard developed by the Darwin Tree of Life Project:
prefix ``m'' stands for mammals,
``b'' for birds,
``a'' for \emph{Amphibia} (e.g. frogs and newts),
``f'' for fish,
``s'' for sharks,
``ic'' for order \emph{Coleoptera} (beetles),
``id'' for \emph{Diptera} (flies),
``ih'' for \emph{Hemiptera} (true bugs),
``il'' for \emph{Lepidoptera} (butterflies and moths),
``iy'' for \emph{Hymenoptera} (bees and ants).
Species marked by ``*'' are used for training cross-species models.
\end{tablenotes}
\end{table}

\subsection{Datasets}

We acquired the genome sequences and gene annotations for six model organisms and 16 non-model organisms (Table~\ref{tab:data}).
For model organisms, only chicken is annotated with the Ensembl pipeline;
others are annotated by third parties.
For non-model organisms, we intentionally chose species that have PacBio HiFi assemblies
and are annotated by both RefSeq and Ensembl if possible.
Only aDenEbr and sMobHyp do not have Ensembl annotations.

\subsection{Intra-species training}

Gencode provides a smaller set of ``basic'' gene annotations and a larger set of ``comprehensive'' annotations.
We also have an option to select the longest transcript of each protein-coding gene for high-fidelity splice sites.
We found training data has minor effect on testing accuracy (Fig.~\ref{fig:3}a)
but annotations used for testing has a larger effect (Fig.~\ref{fig:3}b).
We decided to train on splice sites from the longest protein-coding transcripts for training as they are most accurate,
and to test on basic annotations because non-model organisms probably do not have annotations comparable to comprehensive Gencode anntations.

\begin{figure}[tb]
\includegraphics[width=.48\textwidth]{fig3}
\caption{Intra-species training.
{\bf (a)} Models trained on different human annotations
and tested on the Gencode basic annotation.
{\bf (b)} Model trained from the longest human protein-coding transcripts
and tested on different human annotations.
{\bf (c)} Donor-only training versus joint donor-acceptor training.
{\bf (d)} Acceptor-only training versus joint donor-acceptor training.}\label{fig:3}
\end{figure}

The two central bases in donor training data are always {\tt GT}
and the two bases in acceptor training data are always {\tt AG}.
We speculated 1D-CNN models could easily learn the difference, so we mixed donor and acceptor training data
and trained one joint model for each species.
The joint models achieve nearly the same accuracy as separate donor or acceptor models (Fig.~\ref{fig:3}c,d).
In later experiements, we thus always train one joint model to simplify the training process.

\subsection{Cross-species training}

Our end goal is to improve spliced alignment accuracy for species without known annotations.
Training and prediction are often applied to different species.
To test how well a model trained from one species can predict splice sites in a different species,
we applied models trained from model organisms to human (Fig.~\ref{fig:4}a).
We can see the test accuracy drops quickly with increased evolution distance.
The mouse model is almost good as the human model because mammals are closely related.
We also applied the mosquito model to the fruitfly genome (Fig.~\ref{fig:4}b).
The test accuracy is lower than the accuracy we obtain with the fruitfly model,
but the drop is much smaller in comparison to applying the mosquito model to human.
These experiments suggested we can achieve reasonable accuracy with a model trained from a closely related species.

\begin{figure}[bt]
\includegraphics[width=.48\textwidth]{fig4}
\caption{Cross-species training.
{\bf (a)} Models trained from different species and tested on human.
{\bf (b)} Intra- versus cross-species training with mosquito and fruitfly.}\label{fig:4}
\end{figure}

\begin{figure}[tb]
\includegraphics[width=.48\textwidth]{fig5}
\caption{Training from multiple species.
{\bf (a)} Accuracy of model vi2 on vertebrate genomes. vi2 is trained from six verberate and seven insect genomes.
{\bf (b)} Accuracy of vi2 on insect genomes. Odd chromosomes in starred species are used for training.
Testing is applied to even chromosomes only.
{\bf (c)} Comparison between vi2 and a model trained from verbebrate genomes (v2).
{\bf (d)} Comparison between vi2 and a model trained from insect genomes (i2).}\label{fig:5}
\end{figure}

\begin{figure}[bt]
\includegraphics[width=.48\textwidth]{fig6}
\caption{Effect of window size.
{\bf (a)} Comparison between the default 202bp window size ($\pm$100bp around {\tt GT} or {\tt AG}) and 102bp.
{\bf (b)} Comparison between 202bp and 302bp window sizes.
{\bf (c)} Activation differences between positive and negative donor sites at the last max-pooling layer.
When the window size is set to 302bp, this layer consists of $16\times32$ non-negative cells.
Given a set of sequences, the activation rate of a cell is the frequency of the cell being positive (i.e. non-zero).
The heat shows the rate difference between positive and negative sets.
Dark red indicates higher activation among positive donor sites; dark blue is the opposite.
{\bf (d)} Activation difference for acceptor sites.}\label{fig:6}
\end{figure}

We went a step further by combining the training data across multiple vertebrate and insect species
and derived model vi2, which was trained from six vertebrate and seven insect genomes marked in Table~\ref{tab:data}.
Although this model is not as accurate as the model trained from individual species itself (Fig.~\ref{fig:5}a,b),
it is better than applying an insect model to human (Fig.~\ref{fig:4}a).
Note that vi2 is not trained on amphibian or shark genomes but it still accurately predicts splice sites in aDenEbr and sMobHyp.
It is capturing common signals across large evolutionary distance while reducing overfitting to indivdual species.
We also trained a vertebrate-only model (v2) and an insect-only model (i2).
They outperformed vi2 for vertebrate and insect genomes, respectively (Fig.~\ref{fig:5}c,d),
but we deemed the improvement is marginal and outweighed by the convenience
of having one model across vertebrate and insect genomes.

\begin{figure}[bt]
\includegraphics[width=.48\textwidth]{umap}
\caption{UMAP with values at the last max-pooling layer.
Training data was downsampled to 20k splice sites with 10k positive and 10k negative sites.
}\label{fig:umap}
\end{figure}

Model vi2 considers 202bp sequences around splice sites.
We tried to reduce the window size to 102bp but the accuracy noticeably dropped (Fig.~\ref{fig:6}a).
Increasing the window size to 302bp, on the other hand, only had minor effect (Fig.~\ref{fig:6}b).

To understand what signals the model is learning, we visualized the activation rates at the last max-pooling layer which consists of 16 features (Fig.~\ref{fig:6}c,d).
While some features such as 4 and 7 focus on signals at splice sites,
other features such as 2 and 13 likely capture compositions of introns and exons.
The activation difference fades away beyond 100bp around splice sites especially in introns.
This might explain why increasing window size from 202bp to 302bp has minor effect (Fig.~\ref{fig:6}b).

We further used the values at the last max-pooling layer to generate the UMAP of a subset of training samples (Fig.~\ref{fig:umap}).
The UMAP separtes the training samples into three large clusters which correspond to
positive acceptor sites, negative sites and positive donor sites, respectively.
Notably, mammals and birds contain sites absent from fish and insect.
The UMAP for human, zebrafish and fruitly reveals a similar pattern.
We then extracted splice sites in the UMAP that are specific to human.
We found most of the intron sequences close to the splice sites tend to be short and GC-rich.
We randomly selected five human genes with GC-rich donor sites, including \emph{GAPDH}, \emph{ARHGEF19}, \emph{ABTB2}, \emph{CCAR2} and \emph{OGDH},
and aligned their protein sequences to the zebrafish genome.
Except \emph{CCAR2} which was aligned with low identity at 33\%,
other genes all reserved the GC-rich junctions, though the zebrafish introns had lower GC.
\citet{Zhang:2011aa} also observed GC-rich introns in mammals
and suggested they may help stabilize RNA secondary structure.

\subsection{Improving spliced alignment}

\begin{table*}[!tb]
\caption{Effect of splice site models on protein-to-genome alignment\label{tab:eval}}
\begin{tabular*}{\textwidth}{@{\extracolsep\fill}llllllclllll@{\extracolsep\fill}}
\toprule
& \multicolumn{5}{c}{zebrafish proteins to human genome} && \multicolumn{5}{c}{mosquito proteins to fruitfly genome} \\
\cline{2-6}\cline{8-12}
Aligner           & miniprot  & miniprot  & miniprot  & miniprot   & Spaln3    && miniprot & miniprot & miniprot & miniprot   & Spaln3 \\
Splice model      & GT-AG     & default   & -j2       & minisplice & Tetrapod  && GT-AG    & default  & -j2      & minisplice & InsectDm \\
\midrule
\# prediced junc  & 168,030   & 164,094   & 160,935   & 164,860    & 144,544   && 30,279   & 28,780   & 27,307   & 28,722     & 24,538 \\
\# annotated junc & 144,495   & 146,898   & 147,880   & 157,654    & 132,394   && 24,022   & 24,465   & 24,203   & 27,107     & 21,161 \\
\% unannotated    & 14.01     & 10.48     & 8.11      & 4.37       & 8.41      && 20.68    & 14.99    & 11.37    & 5.62       & 13.76 \\
\% Base Sn        & 60.09     & 60.12     & 60.04     & 60.85      & 51.71     && 57.00    & 56.92    & 56.76    & 57.28      & 44.53 \\
\% Base Sp        & 94.33     & 95.49     & 96.09     & 97.30      & 94.37     && 98.39    & 98.65    & 98.64    & 99.28      & 97.37 \\
\botrule
\end{tabular*}
\begin{tablenotes}\setlength\itemsep{0.0em}
For the splice model, ``GT-AG'' only considers {\tt GT..AG},
``default'' considers {\tt GTR..YAG},
``-j2'' additionally considers {\tt G|GTR..YYYAG},
and ``minisplice'' uses the ``vi2-7k'' model on top of ``-j2''.
Miniprot was invoked with ``-I'' which sets the maximum intron length to $\min\{\max\{3.6\sqrt{G},10^4\},3\times10^5\}$, where $G$ is the genome size.
Spaln3 was invoked with ``-LS -yS -yB -yZ -yX2'' and with its own splice models.
\end{tablenotes}
\end{table*}

\section*{Acknowledgments}

\section*{Author contributions}

H.L. conceived the project.
S.Y., N.H. and H.L. implemented the algorithms and analyzed the data.
S.Y. and H.L. drafted the manuscript.

\section*{Conflict of interest}

None declared.

\section*{Funding}

This work is supported by National Institute of Health grant R01HG010040 (to H.L.).

\section*{Data availability}

The minisplice source code is available at \url{https://github.com/lh3/minisplice}.
Pretrained models can be obtained from \url{https://doi.org/10.5281/zenodo.15446314}.

\bibliographystyle{apalike}
{\sffamily\small
\bibliography{minisplice}}

\end{document}
